<?xml version="1.0"?>
<robot name="/home/leanhchien/webots_ws/src/practice_ros/webots_controller/models/pioneer3at.urdf" xmlns:xacro="http://ros.org/wiki/xacro">
  <material name="blue">
    <color rgba="0 0 0.8 1"/>
  </material>
  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>
  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>  
  <material name="red">
    <color rgba="1 0 0 1"/>
  </material>  
  <material name="green">
    <color rgba="0 1 0 1"/>
  </material> 
  <!-- ****************** ROBOT WORLD ***************************  -->
  <!-- Define the center of the main robot chassis projected on the ground -->	
  <link name="odom"/>

  <!-- The base footprint of the robot is located underneath the chassis -->
  <joint name="base_odom" type="fixed">
    <parent link="odom"/>
    <child link="base_footprint" />
    <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
  </joint>
  <link name="base_footprint"/>

  <!-- The base footprint of the robot is located underneath the chassis -->
  <joint name="base_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link" />
    <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
  </joint>
  <link name="base_link">
    <visual>
      <origin xyz="0 0 0.173" rpy="1.5708 0.00001 -1.5708"/>
      <geometry>
        <box size="0.197 0.2 0.402"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.173" rpy="1.5708 0.00001 -1.5708"/>
      <geometry>
        <box size="0.197 0.2 0.402"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0.173" rpy="1.5708 0.00001 -1.5708"/>
      <geometry>
        <box size="0.264 0.2 0.332"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.173" rpy="1.5708 0.00001 -1.5708"/>
      <geometry>
        <box size="0.264 0.2 0.332"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.064 0 0.273" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.19" length="0.004"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.064 0 0.273" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.19" length="0.004"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.0835 0 0.273" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.17" length="0.004"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0835 0 0.273" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.17" length="0.004"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.106 0 0.247" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.145" length="0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.106 0 0.247" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.145" length="0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.104 0 0.247" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.145" length="0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.104 0 0.247" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.145" length="0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.0167 0.0828 0.284" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.0058" length="0.016"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0167 0.0828 0.284" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.0058" length="0.016"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.017 -0.0855 0.284" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.0058" length="0.016"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.017 -0.0855 0.284" rpy="0 0 -1.5708"/>
      <geometry>
        <cylinder radius="0.0058" length="0.016"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.1758 0.097 0.148" rpy="1.570806 0.000004 -0.7508"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.1758 0.097 0.148" rpy="1.570806 0.000004 -0.7508"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.1758 -0.097 0.148" rpy="1.57079 0.00001 -2.3908"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.1758 -0.097 0.148" rpy="1.57079 0.00001 -2.3908"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.1558 -0.097 0.148" rpy="1.570806 0.000004 -0.7508"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.1558 -0.097 0.148" rpy="1.570806 0.000004 -0.7508"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.1558 0.098 0.148" rpy="1.57079 0.00001 -2.3908"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.1558 0.098 0.148" rpy="1.57079 0.00001 -2.3908"/>
      <geometry>
        <box size="0.049 0.15 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.134 0 0.111" rpy="1.553436 -1.570793 -3.124232"/>
      <geometry>
        <cylinder radius="0.025" length="0.4"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.134 0 0.111" rpy="1.553436 -1.570793 -3.124232"/>
      <geometry>
        <cylinder radius="0.025" length="0.4"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.132 0 0.111" rpy="1.553436 -1.570793 -3.124232"/>
      <geometry>
        <cylinder radius="0.025" length="0.4"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.132 0 0.111" rpy="1.553436 -1.570793 -3.124232"/>
      <geometry>
        <cylinder radius="0.025" length="0.4"/>
      </geometry>
    </collision>
  </link>
 
  <joint name="front_right_motor" type="continuous">
    <parent link="base_link"/>
    <child link="front right wheel"/>
    <axis xyz="-1 -0.00001 0.000003"/>
    <limit effort="20" velocity="6.4"/>
    <origin xyz="0.1331 -0.197 0.11" rpy="1.5708 0.00001 -1.5708"/>
  </joint>
  <link name="front right wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
    </collision>
  </link>
  <joint name="front_left_motor" type="continuous">
    <parent link="base_link"/>
    <child link="front left wheel"/>
    <axis xyz="-1 -0.00001 0.000003"/>
    <limit effort="20" velocity="6.4"/>
    <origin xyz="0.1331 0.197 0.11" rpy="1.5708 0.00001 -1.5708"/>
  </joint>
  <link name="front left wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
    </collision>
  </link>
  <joint name="back_right_motor" type="continuous">
    <parent link="base_link"/>
    <child link="back right wheel"/>
    <axis xyz="-1 -0.00001 0.000003"/>
    <limit effort="20" velocity="6.4"/>
    <origin xyz="-0.1331 -0.197 0.11" rpy="1.5708 0.00001 -1.5708"/>
  </joint>
  <link name="back right wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
    </collision>
  </link>
  <joint name="back_left_motor" type="continuous">
    <parent link="base_link"/>
    <child link="back left wheel"/>
    <axis xyz="-1 -0.00001 0.000003"/>
    <limit effort="20" velocity="6.4"/>
    <origin xyz="-0.1331 0.197 0.11" rpy="1.5708 0.00001 -1.5708"/>
  </joint>
  <link name="back left wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.570796 0.000001 1.570799"/>
      <geometry>
        <cylinder radius="0.11" length="0.095"/>
      </geometry>
    </collision>
  </link>
  <link name="imu">
    <visual>
      <geometry>
        <cylinder radius="0.02" length="0.01"/>
      </geometry>
      <material name="green"/>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="0.02" length="0.01"/>
      </geometry>
    </collision>
  </link>
  <joint name="base_link_imu_joint" type="fixed">
    <parent link="base_link"/>
    <child link="imu"/>
    <origin xyz="0.08 0 0.28" rpy="0 0 0"/>
  </joint>
  <link name="solid">
  </link>
  <joint name="imu_solid_joint" type="fixed">
    <parent link="imu"/>
    <child link="solid"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  <link name="lidar">
    <visual>
      <origin xyz="0.014 0 -0.014" rpy="1.5708 0.00001 -1.5708"/>
      <geometry>
        <box size="0.07 0.01 0.1"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0.014 0 -0.014" rpy="1.5708 0.00001 -1.5708"/>
      <geometry>
        <box size="0.07 0.01 0.1"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.032" length="0.021"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.032" length="0.021"/>
      </geometry>
    </collision>
  </link>
  <joint name="base_link_lidar_joint" type="fixed">
    <parent link="base_link"/>
    <child link="lidar"/>
    <origin xyz="0 0 0.3" rpy="0 0 0"/>
  </joint>
  <link name="screw3">
    <visual>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </collision>
  </link>
  <joint name="lidar_screw3_joint" type="fixed">
    <parent link="lidar"/>
    <child link="screw3"/>
    <origin xyz="0.035 0.0245 -0.0105" rpy="0 1.570787 0"/>
  </joint>
  <link name="screw2">
    <visual>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </collision>
  </link>
  <joint name="lidar_screw2_joint" type="fixed">
    <parent link="lidar"/>
    <child link="screw2"/>
    <origin xyz="0.035 0.0245 -0.0105" rpy="0 1.570787 0"/>
  </joint>
  <link name="screw1">
    <visual>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </collision>
  </link>
  <joint name="lidar_screw1_joint" type="fixed">
    <parent link="lidar"/>
    <child link="screw1"/>
    <origin xyz="0.035 0.0245 -0.0105" rpy="0 1.570787 0"/>
  </joint>
  <link name="screw0">
    <visual>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0.01 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0015" length="0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0015 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.0025" length="0.003"/>
      </geometry>
    </collision>
  </link>
  <joint name="lidar_screw0_joint" type="fixed">
    <parent link="lidar"/>
    <child link="screw0"/>
    <origin xyz="0.035 0.0245 -0.0105" rpy="0 1.570787 0"/>
  </joint>
  <joint name="lidar_secondary_motor" type="continuous">
    <parent link="lidar"/>
    <child link="solid_0"/>
    <axis xyz="0 -0.000004 1"/>
    <limit effort="10" velocity="150"/>
    <origin xyz="-0.046 0.01 -0.01" rpy="-0.000014 0.000005 3.141587"/>
  </joint>
  <link name="solid_0">
  </link>
  <joint name="lidar_main_motor" type="continuous">
    <parent link="lidar"/>
    <child link="lower"/>
    <axis xyz="0 -0.000004 1"/>
    <limit effort="10" velocity="40"/>
    <origin xyz="-0.046 0.02 -0.01" rpy="-0.000014 0.000005 3.141587"/>
  </joint>
  <link name="lower">
  </link>
</robot>
